### PR TITLE
Fix issues when uploading real pathology data

### LINKED ISSUE #11: Diagnosis TR Table

### LINKED ISSUE #220: Update Pathology Report schema for real data
## Pathology Report
- Missing Histologies or needed spaces
   - RON, favor ChRCC
   - RON, favor RO
   - Sarcomatoid, NOS
   - Nephroblastoma (Wilms tumor)
   - Leiomyosarcoma 
   - Renal Cell Carcinoma, NOS
   - Sarcomatoid RCC
   - Unclassified RCC
   - Poorly differentiated malignancy

- TNM Stage
 - Not available

- ajcc_p_stage
  - 1

- pT Stage 
  - Not applicable
  - Not available

- pN Stage
  - Not applicable

- TNM Stage
  - Not applicable

- Focality
  - Not available

- report_source
  - Changed spelling to be "Tumor Registry"

- Margins
  -  Not available
  - Not applicable

- Remove most dependencies
  - More explicit Sacramtoid dependency, with percentage needing to have a status

- tumor_size_range
  - If blank and nephrectomy then do unknown

## Surgery Procedure
Add the following procedures
- Reamings
- Not available

- Nephrectomy Details
  - Remove Robotic assist as required
  - Add to approach and type
    - Not available

### LINKED ISSUE #225: Vital status typo
Typo: "Alived or deceased"

Should be "Alive"

### LINKED ISSUE #226: Multiple radiation treatments on same site is collapsed
Multiple treatments to same site with different start times are collapsed. into single point.

### LINKED ISSUE #234: Simplify tumor_size_range code
Simplify the tumor_size_range code to not be redudant.

use embedding and only calculate at the pathology_report level.

### LINKED ISSUE #235: Fix logic of Diagnosis Date with cases without nephrectommy
Take the earliest date between nephrectomy and mets as defined by (path or metasis by radiation data)

Switch to pathology report, from surgery procedure
* "Biopy"
* "Metasis"

### PR DESCRIPTION
**Pull request by vsmalladi**
Fixes issues that I noticed when loading real pathology data.

- Closes #220 Real Pathology data updates

- Closes #225 Fix typo in vital status

- Closes #226 Fix collapsing of multiple radiation treatment on a single site

- Closes #234 Simply tumor_range code to be only at path reports

- Closes #235 Fix logic of diagnosis date to do any mets and nephrectomy first then other data if not available.

- Closes #11 Tumor Registry Diagnosis date

### CODE REVIEW

**Comment by SenseCode** [comment]
Cannot load the following two:
image
image

**Comment by SenseCode** [comment]
Diagnosis date filtered with surgery and medication seems have problems. Surgery type cannot show up until zoom out. 
image
Medication diagnosis date is not right.
image
Some vital results do not have range. 
image

**Comment by SenseCode** [comment]
Should we put a `first date of visit`, comparing to `last date of follow-up`? Then in the middle put `diagnosis of date`.

**Comment by mingjiecn** [comment]
For calculated diagnosis date, here is a few thoughts:
1.  if diagnosis date is not available for most patient, we need to have a date of first treatment. This date will be the calculated base of all the treatments including surgery, radiation and medication. And this date will be used to set the range of all the treatment chart. 
2. For diagnosis date, we need to consult with Payal to get the logic right. What data need to be included(right now our logic includes surgery with path report, radiation and mediacation). And how to decide the diagnosis date from those data.

**Comment by vsmalladi** [comment]
> For calculated diagnosis date, here is a few thoughts:
> 
> 1. if diagnosis date is not available for most patient, we need to have a date of first treatment. This date will be the calculated base of all the treatments including surgery, radiation and medication. And this date will be used to set the range of all the treatment chart.

Agreed Will create new date of first treatment

> 2. For diagnosis date, we need to consult with Payal to get the logic right. What data need to be included(right now our logic includes surgery with path report, radiation and mediacation). And how to decide the diagnosis date from those data.

* Nephrectomy or biopsy path data
* Medication data
* Tumor Registry data

**Comment by SenseCode** [comment]
Great, all data can be loaded. PR part tumor_size seems have problems as follow: 
image
image
Age older than 90 shown now, not sure we should fix in this branch or not..
image

Great jobs.

File: src/encoded/types/patient.py [hunk 1]
```diff
@@ -466,51 +475,62 @@ def surgery_summary(self, request, surgery=None):
...
+                nephrectomy_dates.sort(key = lambda date: datetime.strptime(date, '%Y-%m-%d'))
+                diagnosis_date = nephrectomy_dates[0]
+                surgery_dates.append(nephrectomy_dates[0])
+                diagnosis_source = "Pathology Report"
+
+        elif len(non_nephrectomy_dates) > 0:
+            non_nephrectomy_dates.sort(key = lambda date: datetime.strptime(date, '%Y-%m-%d'))
+            surgery_dates.append(non_nephrectomy_dates[0])
+            surgery_dates.sort(key = lambda date: datetime.strptime(date, '%Y-%m-%d'))
+            diagnosis_date = surgery_dates[0]
```
Comment (mingjiecn) [inline src/encoded/types/patient.py:514]: surgery_dates is not needed, diagnosis_date get be calculated by just using nephrectomy_dates and non_nephrectomy_dates.

File: src/encoded/types/patient.py [hunk 2]
```diff
@@ -466,51 +475,62 @@ def surgery_summary(self, request, surgery=None):
...
+            diagnosis_source = "Pathology Report"
+
+        elif diagnosis_date_tumor_registry is not None:
+             diagnosis_date = diagnosis_date_tumor_registry
+        elif len(medication) > 0:
+            diagnosis_source = "Medication"
+            for medication_record in medication:
+                medication_object = request.embed(medication_record, '@@object')
+                non_nephrectomy_dates.append(medication_object['start_date'])
                 non_nephrectomy_dates.sort(key = lambda date: datetime.strptime(date, '%Y-%m-%d'))
```
Comment (mingjiecn) [inline src/encoded/types/patient.py]: sort function should happen outside for loop, after all the medication dates are added into list "non_nephrectomy_dates".

File: src/encoded/types/patient.py [hunk 3]
```diff
@@ -466,51 +475,62 @@ def surgery_summary(self, request, surgery=None):
...
             end_date = datetime.strptime(diagnosis_date, "%Y-%m-%d")
             age = end_date.year - birth_date.year -  ((end_date.month, end_date.day) < (birth_date.month, birth_date.day))
             ageString = str(age)
+
             if age >= 90:
                 ageString = "90 or above"
 
-
             if age >= 80:
                 age_range = "80+"
```
Comment (mingjiecn) [inline src/encoded/types/patient.py:543]: this range "80+" should be "80 - 89" if there is a range of "90 or above"

**Comment by mingjiecn** [comment]
Do we need to add radiation data and surgery with no path data back for diagnosis calculation?

File: src/encoded/types/patient.py (hunk 3 above)
Comment (vsmalladi) [reply src/encoded/types/patient.py:543]: remove the 90 or above code

**Comment by SenseCode** [comment]
Checked the diagnosis_date if this is the final version. Some issues here:
1. Lost diagnosis_date in surgery chart : Patient 398,139,186,692.
image

2. No data shown in with nephrectomy and ablation in surgery data (data showed after zoom in): Patient 396,756,021,359
image

File: src/encoded/types/patient.py (hunk 1 above)
Comment (vsmalladi) [reply src/encoded/types/patient.py:514]: It just to find the earliest date between nephrectomy and biopsy. So made an intermediate surgery date to sort those.

Adding comment to clarify

File: src/encoded/types/patient.py (hunk 2 above)
Comment (vsmalladi) [reply src/encoded/types/patient.py]: Fixed this

File: src/encoded/types/patient.py (hunk 3 above)
Comment (vsmalladi) [reply src/encoded/types/patient.py:543]: Clarrifying the reason to have ageString. The 80+ is previously what Payal wanted #50

**Comment by vsmalladi** [comment]
> Checked the diagnosis_date if this is the final version. Some issues here:
> 
> 1. Lost diagnosis_date in surgery chart : Patient 398,139,186,692.
>    image
> 2. No data shown in with nephrectomy and ablation in surgery data (data showed after zoom in): Patient 396,756,021,359
>    image

The diagnosis-date is calculated only with pathology reports. So all of these patients have certain issues.

1. Has surgery_procedure as nephrectomy but no pathology_report so no diagnosis

2. These seem fine Rebuild and check. Example patient: http://localhost:8000/patients/KCEPT021XWS/

Screen Shot 2020-08-11 at 5 35 42 PM

**Comment by vsmalladi** [comment]
> Great, all data can be loaded. PR part tumor_size seems have problems as follow:
> image
> image

Only faceting on pathology reports that are nephrectomy as the biopsy and metasis data don't have tumor_size as known information. So this is more of an issue with the test data.

**Comment by SenseCode** [comment]
Seems we need to load with real data to check if our logic works fine or not.
 If a surgery chart shows nothing in a chart range, that will be weird for the user, and we might need to clarify what's happening in this situation. 
I think in real data, some of the nephrectomy should have `path_report` to confirm it's a benign tumor or not, so should be related to the` diagnosis_date` calculation.

Great jobs.

**Comment by SenseCode** [comment]
For tumor_size range facets, I think it might be my previous codes have bug. Fixed it now, all tumor_size in PR showing right with all right numbers. Hide no value parts when do searching. 

image

**Review by SenseCode — APPROVED** [review]
Overall looks great. Can merge after check the PR part.

File: src/encoded/types/patient.py (hunk 1 above)
Comment (mingjiecn) [reply src/encoded/types/patient.py:514]: should use if if to check nephrectomy_dates and non_nephrectomy_dates, instead of if elif

**Review by mingjiecn — APPROVED** [review]

**Pull request merged**